vault/ui_custom_messages/manager.go

```diff
@@ -0,0 +1,236 @@
+func NewManager(storage logical.Storage) *Manager {
+	return &Manager{
+		view: storage,
+	}
+}
+
+// FindMessages handles getting a list of existing messages that match the
+// criteria set in the provided FindFilter struct.
+func (m *Manager) FindMessages(ctx context.Context, filters FindFilter) ([]Message, error) {
+	nsList, err := getNamespacesToSearch(ctx, filters)
+	if err != nil {
+		return nil, err
+	}
+
+	results := make([]Message, 0)
+
+	for _, ns := range nsList {
+		entry, err := m.getEntryForNamespace(ctx, ns)
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, entry.FindMessages(filters)...)
+	}
+
+	return results, nil
+}
+
+// CreateMessage handles adding the provided Message in the current namespace.
+// A ID for the message is automatically generated and the ID field of the
+// returned Message struct is set to this value. If the maximum number of
+// messages already exists, then the message is not added.
+func (m *Manager) CreateMessage(ctx context.Context, message Message) (*Message, error) {
+	m.l.Lock()
+	defer m.l.Unlock()
+
+	entry, err := m.getEntry(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	err = entry.CreateMessage(&message)
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.putEntry(ctx, entry)
+	if err != nil {
+		return nil, err
+	}
+
+	return &message, nil
+}
+
+// ReadMessage handles retrieving the properties of the message in the current
+// namespace specified by the id value from the logical.Storage.
+func (m *Manager) ReadMessage(ctx context.Context, id string) (*Message, error) {
+	entry, err := m.getEntry(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	message, ok := entry.Messages[id]
+	if !ok {
+		return nil, nil
+	}
+
+	return &message, nil
+}
+
+// UpdateMessage handles updating the message referenced by the provided
+// Message struct in the current namespace with its content in the
+// logical.Storage.
+func (m *Manager) UpdateMessage(ctx context.Context, message Message) (*Message, error) {
+	m.l.Lock()
+	defer m.l.Unlock()
+
+	entry, err := m.getEntry(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	found, err := entry.UpdateMessage(&message)
+	if err != nil {
+		return nil, err
+	}
+
+	if !found {
+		// If no error occured but the specified message doesn't exist...
+		return nil, nil
+	}
+
+	if err = m.putEntry(ctx, entry); err != nil {
+		return nil, err
+	}
+
+	return &message, nil
+}
+
+// DeleteMessage handles deleting the message with the provided id value in the
+// current namespace if it exists. The method updates the logical.Storage as
+// well.
+func (m *Manager) DeleteMessage(ctx context.Context, id string) error {
+	m.l.Lock()
+	defer m.l.Unlock()
+
+	entry, err := m.getEntry(ctx)
+	if err != nil {
+		return err
+	}
+
+	delete(entry.Messages, id)
+
+	return m.putEntry(ctx, entry)
+}
+
+// getEntry is a helper method that retrieves the current namespace from the
+// context.Context and uses it to call getEntryForNamespace.
```

Comment:
That's not the case though. A lock is only held when an operation will call putEntry. This protects against another function that also has a putEntry call in it that could race in and commit a conflicting change. But functions that only have a getEntry call, don't need to use the lock.

Reply:
What if I'm trying to do a Get on an entry that's also being updated? I might be missing something obvious, but a read lock intuitively makes sense for a function like this

Reply:
Since the underlying physical.Backend ensures that data returned by a Get call is consistent, it doesn't matter. The unprotected getEntry call here will get the logical.StorageEntry in either its state prior to a concurrent Put or immediately after, but it will not get a half updated state. more than one logical.StorageEntry was being retrieved than a _read lock_ would be needed to make sure that the states of all the entries were consistent with each other.